### PR TITLE
[SYCL][HIP][libclc] Wire up AMD half support

### PR DESCRIPTION
**Pull request by MartinWehking**
Enables fp16 support for AMD GPUs.
Based on Zheming's previous work: [8488](https://github.com/intel/llvm/pull/8488)
Some test cases for e.g. images were disabled since they aren't supported

Draft PR for enabling fp16 support in the unified-runtime: https://github.com/oneapi-src/unified-runtime/pull/988

### CODE REVIEW

**Comment by jinz2014** [comment]
Thank you for your extensions and support !

**Comment by JackAKirk** [comment]
Looks good, I'll wait for the CI testing though.

**Comment by ldrumm** [comment]
Our current squash policy is going to mangle authorship of this patchset. What are we going to do about it

**Review by dm-vodopyanov — APPROVED** [review]

**Comment by JackAKirk** [comment]
You can see from e.g. the log from the unexpected passing test, reduction_nd_ext_half.cpp, that the CI run was skipping all the tests requiring `aspect::fp16`. I think that you either need to get https://github.com/oneapi-src/unified-runtime/pull/988/ merged first, or if there is a way of pointing the CI to that unified-runtime PR branch do that. Otherwise we can't see that any of the affected tests are passing on the CI.

File: sycl/test-e2e/Reduction/reduction_nd_ext_half.cpp
Comment (JackAKirk) [inline sycl/test-e2e/Reduction/reduction_nd_ext_half.cpp]: I don't think it makes sense to remove XFAIL when it was only unexpectedly passing because it was being skipped because the CI device didn't have the fp16 aspect: it will have the aspect once you merge the unified-runtime patch.

Comment (JackAKirk) [reply sycl/test-e2e/Reduction/reduction_nd_ext_half.cpp]: Yeah that makes sense

File: libclc/amdgcn-amdhsa/libspirv/math/lgamma.cl
Comment (JackAKirk) [inline libclc/amdgcn-amdhsa/libspirv/math/lgamma.cl:19]: probably a nit but I don't understand this: why enable an OPENCL EXTENSION on a non opencl backend? Presumably the `cl_khr_fp64` was already working before this patch so I don't imagine this changes any behaviour.

Comment (MartinWehking) [reply libclc/amdgcn-amdhsa/libspirv/math/lgamma.cl:19]: This has been resolved in internal discussions already, but I'd still like to post it here as it might be interesting for others as well.
The fact that the AMDGPU target is a non-opencl doesn't matter here.
We basically enable the OpenCL  fp64 extension because libclc is compiled by an OpenCL C compiler (i.e. it needs to be compliant with the openCL language specification).
Not enabling it would lead to the fact that a fp64 literal declared inside this cl file could be compiled as a float.
This is basically where Vanilla OpenCL C deviates from the C standard.
Potentially, this could lead to unexpected behaviour in the future and therefore, it's safer to enable the extension.

For the fp16 extension that is also enabled several times in cl files by this patch, it's a different story.
Not enabling it (i.e. removing the pragmas) leads to errors at compile time with the error message that the `half` type is not allowed.
It's needed to actually make it a type:
https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#reinterpreting-types-using-as_type-and-as_typen
(footnote 15)

This is also why @jinz2014 added these pragmas and wrapped them inside the ifdef macro checks

**Comment by MartinWehking** [comment]
ping @intel/llvm-gatekeepers to get this merged

**Comment by JackAKirk** [comment]
> ping @intel/llvm-gatekeepers to get this merged

I think this needs a review from @intel/dpcpp-l0-pi-reviewers first

**Comment by smaslov-intel** [comment]
I don't see any changes to the files owned by @intel/dpcpp-l0-pi-reviewers , but I gave my provisional approval (without a review)

**Review by smaslov-intel — APPROVED** [review]
LGTM

**Comment by MartinWehking** [comment]
@smaslov-intel I had to change the UR vars in `sycl/plugins/unified_runtime/CMakeLists.txt` at some point to make the CI use my UR patch. This automatically requested a review from you.
I reverted those changed again in https://github.com/MartinWehking/llvm/commit/51fe9311e77820564e197295c79247d82fb6cdb8

**Pull request merged**
